InfiniteRecharge/src/test/java/frc/robot/utils/control/statespace/estimators/kalman/KalmanFilterTest.java

```diff
@@ -0,0 +1,127 @@
+        double dt = 0.02;
+
+
+
+        MotorPosition model = new MotorPosition(dt, type, I);
+
+        model.setState(new SimpleMatrix(new double[][] {
+            new double[] {0},
+            new double[] {0}
+        }));
+
+        SimpleMatrix P0 = new SimpleMatrix(new double[][] {
+            new double[] {0.01, 0},
+            new double[] {0, 0.02},
+        });
+
+
+
+        SimpleMatrix R = new SimpleMatrix(new double[][] {
+            new double[] {0.03, 0},
+            new double[] {0, 0.04},
+        });
+
+        SimpleMatrix Q = new SimpleMatrix(new double[][] {
+            new double[] {0.05, 0},
+            new double[] {0, 0.06},
+        });
+
+        SimpleMatrix G = new SimpleMatrix(new double[][] {
+            new double[] {1, 2},
+            new double[] {3, 4}
+        });
+
+        // idk what kind of encoder gives us this but don't judge
+        SimpleMatrix C = new SimpleMatrix(new double[][] {
+            new double[] {2, 3},
+            new double[] {-0.1, 5}
+        });
+
+        SimpleMatrix D = new SimpleMatrix(2, 2);
+
+        KalmanFilter filter = new KalmanFilter(P0, model) {
+            @Override
+            protected SimpleMatrix getR(SimpleMatrix state, SimpleMatrix input, double t) {
+                return R;
+            }
+        
+            @Override
+            protected SimpleMatrix getQ(SimpleMatrix state, SimpleMatrix input, double t) {
+                return Q;
+            }
+        
+            @Override
+            protected SimpleMatrix getG(SimpleMatrix state, SimpleMatrix input, double t) {
+                return G;
+            }
+        
+            @Override
+            protected SimpleMatrix getD(SimpleMatrix state, SimpleMatrix input, double t) {
+                return D;
+            }
+        
+            @Override
+            protected SimpleMatrix getC(SimpleMatrix state, SimpleMatrix input, double t) {
+                return C;
+            }
+        };
+
+        model.setState(new SimpleMatrix(2, 1, true, new double[] {Math.PI, Math.PI / 100}));
+
+
+
+        model.apply(new SimpleMatrix(1, 1, true, new double[] {8}));
+        filter.predict();
+
+        SimpleMatrix y = new SimpleMatrix(2, 1, true, new double[] {181, 295});
+        filter.update(y);
+
+        assertTrue(model.getState().isIdentical(new SimpleMatrix(2, 1, true, new double[] {
+            2.650741871503023, 58.865022680661937
+        }), TOLERANCE));
+
+
+
+        model.apply(new SimpleMatrix(1, 1, true, new double[] {6}));
+        filter.predict();
+
+        y = new SimpleMatrix(2, 1, true, new double[] {295, 470});
+        filter.update(y);
+
+        assertTrue(model.getState().isIdentical(new SimpleMatrix(2, 1, true, new double[] {
+            5.338061309893673, 94.364827040089935
+        }), TOLERANCE));
+
+        double t1 = System.nanoTime();
+
+        System.out.println((t1 - t0) / 1000000 + "ms to run");
```

Comment:
Typically you don't want System.out.println()'s in unit tests. The output should be empty so the test report is clean.

Reply:
oh okay, is there an acceptable way to make sure the calculations don't take too long, preferably showing the time they took somewhere?

Reply:
hmmm.. I guess it depends on how you are using it. Every machine executing the unit test is going to be different, so it's not useful in the output of our build logs. If you are using it as a mental check when you run the test, that's fine.